File: sdks/typescript/src/skills.ts
/**
 * Skill invocation helpers for the Meerkat SDK.
 */

import type { MeerkatClient } from "./client.js";
import type { WireRunResult } from "./generated/types.js";
import { CapabilityUnavailableError } from "./generated/errors.js";

export type SkillKey = {
  source_uuid: string;
  skill_name: string;
};

export type SkillRef = SkillKey | string;

function isStructured(ref: SkillRef): ref is SkillKey {
  return typeof ref !== "string";
}

function normalizeRef(ref: SkillRef): SkillKey {
  if (isStructured(ref)) return ref;
  const value = ref.startsWith("/") ? ref.slice(1) : ref;
  const [source_uuid, ...rest] = value.split("/");
  if (!source_uuid || rest.length === 0) {
    throw new Error(
      `Invalid legacy skill reference '${ref}'. Expected '<source_uuid>/<skill_name>'.`,
    );
  }
  console.warn(
    `[meerkat-sdk] legacy skill reference '${ref}' is deprecated; pass { source_uuid, skill_name } instead.`,
  );
  return { source_uuid, skill_name: rest.join("/") };
}

/**
 * Helpers for working with Meerkat skills.
 */
export class SkillHelper {
  private client: MeerkatClient;

  constructor(client: MeerkatClient) {
    this.client = client;
  }

  isAvailable(): boolean {
    return this.client.hasCapability("skills");
  }

  requireSkills(): void {
    if (!this.isAvailable()) {
      throw new CapabilityUnavailableError(
        "CAPABILITY_UNAVAILABLE",
        "Skills capability is not available in this runtime. " +
          "Build with --features skills to enable.",
      );
    }
  }

  async invoke(
    sessionId: string,
    skillRef: SkillRef,
    prompt: string,
  ): Promise<WireRunResult> {
    this.requireSkills();
    return this.client.startTurn(sessionId, prompt, {
      skill_refs: [normalizeRef(skillRef)],
    });
  }

  async invokeNewSession(
    skillRef: SkillRef,
    prompt: string,
    model?: string,
  ): Promise<WireRunResult> {
    this.requireSkills();
    return this.client.createSession({
      prompt,
      model,
      skill_refs: [normalizeRef(skillRef)],
    });
  }

  async listResources(sessionId: string, skillRef: SkillRef): Promise<WireRunResult> {
    this.requireSkills();
    const canonical = normalizeRef(skillRef);
    const id = `${canonical.source_uuid}/${canonical.skill_name}`;
    return this.client.startTurn(
      sessionId,
      `Use skill_list_resources for id '${id}' and return only the tool result JSON.`,
      { skill_refs: [canonical] },
    );
  }

  async readResource(
    sessionId: string,
    skillRef: SkillRef,
    path: string,
  ): Promise<WireRunResult> {
    this.requireSkills();
    const canonical = normalizeRef(skillRef);
    const id = `${canonical.source_uuid}/${canonical.skill_name}`;
    return this.client.startTurn(
      sessionId,
      `Use skill_read_resource for id '${id}' and path '${path}', then return only the tool result JSON.`,
      { skill_refs: [canonical] },
    );
  }

  async invokeFunction(
    sessionId: string,
    skillRef: SkillRef,
    functionName: string,
    argumentsValue: unknown,
  ): Promise<WireRunResult> {
    this.requireSkills();
    const canonical = normalizeRef(skillRef);
    const id = `${canonical.source_uuid}/${canonical.skill_name}`;
    return this.client.startTurn(
      sessionId,
      `Use skill_invoke_function for id '${id}', function '${functionName}', arguments: ${JSON.stringify(argumentsValue)}. Return only the tool result JSON.`,
      { skill_refs: [canonical] },
    );
  }
}

File: sdks/python/meerkat/skills.py
"""Skill listing and invocation helpers for the Meerkat SDK."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Union
import warnings

from .errors import CapabilityUnavailableError


@dataclass(frozen=True)
class SkillKey:
    source_uuid: str
    skill_name: str


SkillRef = Union[SkillKey, str]


class SkillHelper:
    """Helpers for working with Meerkat skills."""

    def __init__(self, client):
        self._client = client

    def is_available(self) -> bool:
        return self._client.has_capability("skills")

    def require_skills(self) -> None:
        if not self.is_available():
            raise CapabilityUnavailableError(
                "CAPABILITY_UNAVAILABLE",
                "Skills capability is not available in this runtime. "
                "Build with --features skills to enable.",
            )

    @staticmethod
    def _normalize_ref(skill_ref: SkillRef) -> SkillKey:
        if isinstance(skill_ref, SkillKey):
            return skill_ref

        value = skill_ref[1:] if skill_ref.startswith("/") else skill_ref
        parts = value.split("/")
        if len(parts) < 2:
            raise ValueError(
                f"Invalid legacy skill reference '{skill_ref}'. "
                "Expected '<source_uuid>/<skill_name>'."
            )

        warnings.warn(
            "Legacy string skill references are deprecated; pass SkillKey instead.",
            DeprecationWarning,
            stacklevel=3,
        )
        return SkillKey(source_uuid=parts[0], skill_name="/".join(parts[1:]))

    async def invoke(self, session_id: str, skill_ref: SkillRef, prompt: str):
        self.require_skills()
        canonical = self._normalize_ref(skill_ref)
        return await self._client.start_turn(
            session_id,
            prompt,
            skill_refs=[
                {
                    "source_uuid": canonical.source_uuid,
                    "skill_name": canonical.skill_name,
                }
            ],
        )

    async def invoke_new_session(self, skill_ref: SkillRef, prompt: str, model: str | None = None):
        self.require_skills()
        canonical = self._normalize_ref(skill_ref)
        return await self._client.create_session(
            prompt=prompt,
            model=model,
            skill_refs=[
                {
                    "source_uuid": canonical.source_uuid,
                    "skill_name": canonical.skill_name,
                }
            ],
        )
